lib/pbench/test/unit/server/auth/test_auth.py

```diff
@@ -646,7 +656,54 @@ def test_verify_auth_oidc_user_update(
         assert user.roles == ["ROLE"]
         assert user.username == "new_dummy"
 
-    def test_verify_auth_oidc_invalid(self, monkeypatch, rsa_keys, make_logger):
+    def test_verify_auth_oidc_user_admin(
+        self,
+        monkeypatch,
+        server_config: PbenchServerConfig,
+        db_session,
+        rsa_keys,
+        make_logger,
+    ):
+        """Verify we update our internal user database when we get updated user
+        payload from the OIDC token for an existing user."""
+        client_id = "us"
+        token, expected_payload = gen_rsa_token(client_id, rsa_keys["private_key"])
+
+        # Mock the Connection object and generate an OpenIDClient object,
+        # installing it as Auth module's OIDC client.
+        config = mock_connection(
+            monkeypatch, client_id, public_key=rsa_keys["public_key"]
+        )
+        oidc_client = OpenIDClient.construct_oidc_client(config)
+        monkeypatch.setattr(Auth, "oidc_client", oidc_client)
+        server_config._conf.set("pbench-server", "admin-role", "friend,dummy,admin")
+
+        app = Flask("test-verify-auth-oidc-user-update")
+        app.logger = make_logger
+        app.server_config = server_config
+        with app.app_context():
+            user = Auth.verify_auth(token)
+
+        assert user.id == "12345"
+        assert user.roles == ["ADMIN"]
+        assert user.username == "dummy"
+
+        # Generate a new token with a role for the same user
+        token, expected_payload = gen_rsa_token(
+            client_id,
+            rsa_keys["private_key"],
+            username="friend",
```

Comment:
In the comment at line 691, you say "same user"...but there's a different _username_; should the comment say "user ID" (or, perhaps better yet, should the two calls to `gen_rsa_token()` explicitly pass the user ID, rather than implicitly using the default..."code as documentation")?

Reply:
Tracking "user identity" only by UUID is going to be a problem if I try to combine multiple ID providers. But I don't want to just use "username", either, since if you have a Red Hat SSO account named "wscales" that's got ADMIN but we also support another ID provider where you don't, *I* could grab that username and become ADMIN. I *think* I really want to manage multiple identities by email (which most ID providers do or can verify) so that only verified "wscales@redhat.com" users would be given ADMIN. Nikhil removed the code that extracted email from the auth token, but we're still getting it (I verified on the server side, and you can see it in the dashboard's profile).

Since I'm hoping to follow up this PR with the identity merging, I think a lot of this weirdness will go away. I debated dropping the "update username" when I cloned the original test since it's not directly relevant here, but I ended up not doing that.

Reply:
> Tracking "user identity" only by UUID is going to be a problem if I try to combine multiple ID providers.

That's true only if we want to grant the privilege to the user regardless of which ID provider they use (which is not an obvious requirement).

> I don't want to just use "username", either, since if you have a Red Hat SSO account named "wscales" that's got ADMIN but we also support another ID provider where you don't, I could grab that username and become ADMIN.

Indeed.

> I think I really want to manage multiple identities by email

That's great until the user updates their email address.  And, we also get caught in a bind when the "same user" has different email addresses on different IDPs (e.g., because they updated one and haven't updated the other).

> I think a lot of this weirdness will go away.

🎉  However, in the short term, my choice would be to use "username" and only one IDP.  😇